Description: Fix C++23 incomplete type error in floss_dbus_manager.h
 C++23 requires complete type definition before use in std::unique_ptr.
 Move FlossClientBundle class definition before FlossDBusManager class.
 Use 'friend class' syntax for forward-declared friend classes.
Author: Juan Manuel Mendez Rey <vejeta@gmail.com>
Bug-Debian: https://bugs.debian.org/915400
Forwarded: not-needed
Last-Update: 2025-11-25

diff --git a/device/bluetooth/floss/floss_dbus_manager.h b/device/bluetooth/floss/floss_dbus_manager.h
index 1c14f39..96fa5d4 100644
--- a/device/bluetooth/floss/floss_dbus_manager.h
+++ b/device/bluetooth/floss/floss_dbus_manager.h
@@ -47,6 +47,69 @@ class FlossSocketManager;
 class FlossAdminClient;
 #endif  // BUILDFLAG(IS_CHROMEOS)

+// The bundle of all DBus clients used by FlossDBusManager. The bundle is used
+// to control the correct ordering of creation and deletion of clients before
+// shutting down the system bus.
+// NOTE: Moved before FlossDBusManager for C++23 compatibility (complete type
+// required for std::unique_ptr member).
+class DEVICE_BLUETOOTH_EXPORT FlossClientBundle {
+ public:
+  FlossClientBundle(const FlossClientBundle&) = delete;
+  FlossClientBundle& operator=(const FlossClientBundle&) = delete;
+
+  explicit FlossClientBundle(bool use_stubs);
+  ~FlossClientBundle();
+
+  FlossManagerClient* manager_client() { return manager_client_.get(); }
+
+  FlossAdapterClient* adapter_client() { return adapter_client_.get(); }
+
+  FlossGattManagerClient* gatt_manager_client() {
+    return gatt_manager_client_.get();
+  }
+
+  FlossSocketManager* socket_manager() { return socket_manager_.get(); }
+
+  FlossLEScanClient* lescan_client() { return lescan_client_.get(); }
+
+  FlossLoggingClient* logging_client() { return logging_client_.get(); }
+
+  FlossAdvertiserClient* advertiser_client() {
+    return advertiser_client_.get();
+  }
+#if BUILDFLAG(IS_CHROMEOS)
+  FlossAdminClient* admin_client() { return admin_client_.get(); }
+#endif  // BUILDFLAG(IS_CHROMEOS)
+
+  FlossBatteryManagerClient* battery_manager_client() {
+    return battery_manager_client_.get();
+  }
+
+  FlossBluetoothTelephonyClient* bluetooth_telephony_client() {
+    return bluetooth_telephony_client_.get();
+  }
+
+ private:
+  friend class FlossDBusManagerSetter;
+  friend class FlossDBusManager;
+
+  void ResetAdapterClients();
+
+  bool use_stubs_;
+  std::unique_ptr<FlossManagerClient> manager_client_;
+  std::unique_ptr<FlossAdapterClient> adapter_client_;
+  std::unique_ptr<FlossGattManagerClient> gatt_manager_client_;
+  std::unique_ptr<FlossSocketManager> socket_manager_;
+  std::unique_ptr<FlossLEScanClient> lescan_client_;
+  std::unique_ptr<FlossLoggingClient> logging_client_;
+  std::unique_ptr<FlossAdvertiserClient> advertiser_client_;
+  std::unique_ptr<FlossBatteryManagerClient> battery_manager_client_;
+  std::unique_ptr<FlossBluetoothTelephonyClient> bluetooth_telephony_client_;
+#if BUILDFLAG(IS_CHROMEOS)
+  std::unique_ptr<FlossAdminClient> admin_client_;
+#endif  // BUILDFLAG(IS_CHROMEOS)
+};
+
 // FlossDBusManager manages the lifetimes of D-Bus connections and clients. It
 // ensures the proper ordering of shutdowns for the D-Bus thread, connections
 // and clients.
@@ -358,67 +421,6 @@ class DEVICE_BLUETOOTH_EXPORT FlossDBusThreadManager {
   scoped_refptr<dbus::Bus> system_bus_;
 };

-// The bundle of all DBus clients used by FlossDBusManager. The bundle is used
-// to control the correct ordering of creation and deletion of clients before
-// shutting down the system bus.
-class DEVICE_BLUETOOTH_EXPORT FlossClientBundle {
- public:
-  FlossClientBundle(const FlossClientBundle&) = delete;
-  FlossClientBundle& operator=(const FlossClientBundle&) = delete;
-
-  explicit FlossClientBundle(bool use_stubs);
-  ~FlossClientBundle();
-
-  FlossManagerClient* manager_client() { return manager_client_.get(); }
-
-  FlossAdapterClient* adapter_client() { return adapter_client_.get(); }
-
-  FlossGattManagerClient* gatt_manager_client() {
-    return gatt_manager_client_.get();
-  }
-
-  FlossSocketManager* socket_manager() { return socket_manager_.get(); }
-
-  FlossLEScanClient* lescan_client() { return lescan_client_.get(); }
-
-  FlossLoggingClient* logging_client() { return logging_client_.get(); }
-
-  FlossAdvertiserClient* advertiser_client() {
-    return advertiser_client_.get();
-  }
-#if BUILDFLAG(IS_CHROMEOS)
-  FlossAdminClient* admin_client() { return admin_client_.get(); }
-#endif  // BUILDFLAG(IS_CHROMEOS)
-
-  FlossBatteryManagerClient* battery_manager_client() {
-    return battery_manager_client_.get();
-  }
-
-  FlossBluetoothTelephonyClient* bluetooth_telephony_client() {
-    return bluetooth_telephony_client_.get();
-  }
-
- private:
-  friend FlossDBusManagerSetter;
-  friend FlossDBusManager;
-
-  void ResetAdapterClients();
-
-  bool use_stubs_;
-  std::unique_ptr<FlossManagerClient> manager_client_;
-  std::unique_ptr<FlossAdapterClient> adapter_client_;
-  std::unique_ptr<FlossGattManagerClient> gatt_manager_client_;
-  std::unique_ptr<FlossSocketManager> socket_manager_;
-  std::unique_ptr<FlossLEScanClient> lescan_client_;
-  std::unique_ptr<FlossLoggingClient> logging_client_;
-  std::unique_ptr<FlossAdvertiserClient> advertiser_client_;
-  std::unique_ptr<FlossBatteryManagerClient> battery_manager_client_;
-  std::unique_ptr<FlossBluetoothTelephonyClient> bluetooth_telephony_client_;
-#if BUILDFLAG(IS_CHROMEOS)
-  std::unique_ptr<FlossAdminClient> admin_client_;
-#endif  // BUILDFLAG(IS_CHROMEOS)
-};
-
 }  // namespace floss

 #endif  // DEVICE_BLUETOOTH_FLOSS_FLOSS_DBUS_MANAGER_H_
